pyo3-ffi/src/datetime.rs

```diff
@@ -610,20 +611,27 @@ pub unsafe fn PyDateTime_TimeZone_UTC() -> *mut PyObject {
 
 /// Populates the `PyDateTimeAPI` object
 pub unsafe fn PyDateTime_IMPORT() {
-    // PyPy expects the C-API to be initialized via PyDateTime_Import, so trying to use
-    // `PyCapsule_Import` will behave unexpectedly in pypy.
-    #[cfg(PyPy)]
-    let py_datetime_c_api = PyDateTime_Import();
-
-    #[cfg(not(PyPy))]
-    let py_datetime_c_api = {
-        // PyDateTime_CAPSULE_NAME is a macro in C
-        let PyDateTime_CAPSULE_NAME = CString::new("datetime.datetime_CAPI").unwrap();
-
-        PyCapsule_Import(PyDateTime_CAPSULE_NAME.as_ptr(), 1) as *mut PyDateTime_CAPI
-    };
-
-    *PyDateTimeAPI_impl.0.get() = py_datetime_c_api;
+    if !PyDateTimeAPI_impl.once.is_completed() {
+        // PyPy expects the C-API to be initialized via PyDateTime_Import, so trying to use
+        // `PyCapsule_Import` will behave unexpectedly in pypy.
+        #[cfg(PyPy)]
+        let py_datetime_c_api = PyDateTime_Import();
+
+        #[cfg(not(PyPy))]
+        let py_datetime_c_api = {
+            // PyDateTime_CAPSULE_NAME is a macro in C
+            let PyDateTime_CAPSULE_NAME = CString::new("datetime.datetime_CAPI").unwrap();
+
+            PyCapsule_Import(PyDateTime_CAPSULE_NAME.as_ptr(), 1) as *mut PyDateTime_CAPI
+        };
+
```

Comment:
Can `PyDateTime_Import` ever fail? I guess if the return value is null then it might be unhelpful to continue to set the `once`. 🤔

Reply:
So, the docs for `PyCapsule_Import` [imply](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_Import) that it might set an exeption and fail, but then again the definition of the `PyDateTime_IMPORT` macro doesn't do any error checking:
 https://github.com/python/cpython/blob/12eaadc0ad33411bb02945d700b6ed7e758bb188/Include/datetime.h#L199-L200

so I think in practice it can't fail. Also, our `PyDateTime_IMPORT` FFI wrapper doesn't return anything (following CPython...), so we would need to change the API (or I guess tell people they need to call `PyErr_Occurred` after every import call, which would be kind of awful, especially for an issue that CPython itself doesn't seem to worry about.

I commented on an upstream issue about this: https://github.com/python/cpython/issues/122184#issuecomment-2415171567

Reply:
I commented there, I think it can fail. I think it's good enough here to add

```suggestion
if py_datetime_c_api.is_null() {
    return;
}
```

Reply:
ok, easy enough to add the error check